src/main/java/com/couchbase/client/commons/CouchbaseMap.java

```diff
@@ -0,0 +1,310 @@
+package com.couchbase.client.commons;
+
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import com.couchbase.client.core.message.ResponseStatus;
+import com.couchbase.client.core.message.kv.subdoc.multi.Lookup;
+import com.couchbase.client.core.message.kv.subdoc.multi.Mutation;
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.document.JsonDocument;
+import com.couchbase.client.java.document.json.JsonArray;
+import com.couchbase.client.java.document.json.JsonObject;
+import com.couchbase.client.java.document.json.JsonValue;
+import com.couchbase.client.java.error.CASMismatchException;
+import com.couchbase.client.java.error.DocumentAlreadyExistsException;
+import com.couchbase.client.java.error.subdoc.MultiMutationException;
+import com.couchbase.client.java.error.subdoc.PathNotFoundException;
+import com.couchbase.client.java.subdoc.DocumentFragment;
+
+/**
+ * A CouchbaseMap is a {@link Map} backed by a {@link Bucket Couchbase} document (more specifically a
+ * {@link JsonDocument JSON object}).
+ *
+ * Null keys are NOT permitted, and keys are restricted to {@link String}.
+ *
+ * Values in a CouchbaseMap are restricted to the types that a {@link JsonObject JSON objects}
+ * can contain. JSON sub-objects and sub-arrays can be represented as {@link JsonObject} and {@link JsonArray}
+ * respectively.
+ *
+ * @param <V> the type of values in the map (restricted to {@link JsonObject}.
+ */
+public class CouchbaseMap<V> extends AbstractMap<String, V> {
+
+    public static final int MAX_OPTIMISTIC_LOCKING_ATTEMPTS = CouchbaseArrayList.MAX_OPTIMISTIC_LOCKING_ATTEMPTS;
+
+    private final String id;
+    private final Bucket bucket;
+
+    /**
+     * Create a new {@link CouchbaseMap}, backed by the document identified by <code>id</code>
+     * in the given Couchbase <code>bucket</code>. Note that if the document already exists,
+     * its content will be used as initial content for this collection. Otherwise it is created empty.
+     *
+     * @param id the id of the Couchbase document to back the map.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     */
+    public CouchbaseMap(String id, Bucket bucket) {
+        this.id = id;
```

Comment:
Well, the two ctors have different purposes: without data parameter, it will re-use a pre-existing document. With data parameter (including null or empty), it will reconstruct the underlying document from scratch. Null/empty data can then signal the creation of an empty `Map`.